### PR TITLE
Check that shims are called with the correct number of arguments

### LINKED ISSUE #1272: Check number of arguments for shims
Basically all our shims look something like this:
```rust
                let _attr_place = this.deref_operand(args[0])?;
                let addr_place = this.deref_operand(args[1])?;
                let size_place = this.deref_operand(args[2])?;
```
This code will ICE if the user passes fewer than 3 arguments, and it will silently "just work" if the user passes more than 3 arguments. To mitigate this, I propose we add a helper function `check_arg_count` or so that we call for each and every shim (foreign items and intrinsics) to raise an appropriate error (UB seems right) in case the number of arguments is wrong.

(libstd-only shims that check `this.frame().instance` do not need checking.)

### PR DESCRIPTION
**Pull request by toc-the-younger**
Fixes #1272

Some shims don't use all their arguments, I've assumed they are being called in tests with the correct number of arguments here.

### CODE REVIEW

File: src/shims/foreign_items/posix.rs [hunk 1]
```diff
@@ -248,9 +275,10 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 this.write_null(dest)?;
             }
             "pthread_atfork" => {
+                check_arg_count(args, 3)?;
                 let _prepare = this.read_scalar(args[0])?.not_undef()?;
                 let _parent = this.read_scalar(args[1])?.not_undef()?;
-                let _child = this.read_scalar(args[1])?.not_undef()?;
+                let _child = this.read_scalar(args[2])?.not_undef()?;
```
Comment (toc-the-younger) [inline src/shims/foreign_items/posix.rs:281]: This seems to have just been a latent bug.

File: src/shims/intrinsics.rs [hunk 2]
```diff
@@ -367,21 +383,27 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
-            "exact_div" =>
-                this.exact_div(this.read_immediate(args[0])?, this.read_immediate(args[1])?, dest)?,
+            "exact_div" => {
+                check_arg_count(args, 2)?;
+                this.exact_div(this.read_immediate(args[0])?, this.read_immediate(args[1])?, dest)?
+            }
 
-            "forget" => {}
+            "forget" => {
+                check_arg_count(args, 0)?;
```
Comment (toc-the-younger) [inline src/shims/intrinsics.rs:392]: I think this takes 1 arg, but if so it may not be covered by any test.

**Comment by RalfJung** [comment]
Hi @toc-the-younger, welcome and thank you so much for your first PR here!
I'll add reviewing this to my queue, but as you can see there are some older PRs I have to get to first, so please don't be worried when it takes a bit for me tog et back to you.

File: src/shims/foreign_items/posix.rs (hunk 1 above)
Comment (RalfJung) [reply src/shims/foreign_items/posix.rs:281]: Indeed that was probably a typo on my side, good catch.
Now I wonder if we could prevent this through some type-based magic... like `let [prepare, parent, child] = check_arg_count::<3>(args)` or so. In fact, I think libstd has `TryFrom` instances to turn slices into arrays, so maybe we could use that? Something [like this snippet](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=079d641eabe4a45517654c586f2d7773). Maybe with a macro to make it less repetetive, though that may not be worth it.

We can leave that to a future PR though if you want.

File: src/shims/intrinsics.rs (hunk 2 above)
Comment (RalfJung) [reply src/shims/intrinsics.rs:389]: Yes it does take 1 arg, of arbitrary type, that it does nothing with.
Could you add a test (maybe in a separate PR)?

File: src/helpers.rs [hunk 3]
```diff
@@ -472,6 +472,14 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     }
 }
 
+/// Check that the number of args is what we expect.
+pub fn check_arg_count<'tcx>(args: &[OpTy<'tcx, Tag>], n: usize) -> InterpResult<'tcx> {
+    if args.len() != n {
+        throw_ub_format!("incorrect number of arguments, got {}, needed {}", args.len(), n);
```
Comment (RalfJung) [inline src/helpers.rs:478]: It would be good if this could print the name of the function that was called incorrectly.

File: src/shims/foreign_items.rs [hunk 4]
```diff
@@ -163,6 +166,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                     .iter()
                     .find(|cnum| tcx.is_panic_runtime(**cnum))
                     .expect("No panic runtime found!");
+                check_arg_count(args, 1)?;
```
Comment (RalfJung) [inline src/shims/foreign_items.rs:169]: THis is not needed: we go find the proper MIR of the function here and return it, which means the function calling code in the core engine will pass arguments and also validate that their number is right.

File: src/shims/foreign_items.rs [hunk 5]
```diff
@@ -132,13 +133,15 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 // The implementation is provided by the function with the `#[panic_handler]` attribute.
                 "panic_impl" => {
                     this.check_panic_supported()?;
+                    check_arg_count(args, 1)?;
```
Comment (RalfJung) [inline src/shims/foreign_items.rs:136]: This is likewise not needed, because this is not just a shim -- we are finding and returning proper MIR for this call.

File: src/shims/foreign_items/posix.rs [hunk 6]
```diff
@@ -23,40 +24,51 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
             // File related shims
             "open" | "open64" => {
+                check_arg_count(args, 3)?;
                 let result = this.open(args[0], args[1])?;
                 this.write_scalar(Scalar::from_i32(result), dest)?;
             }
             "fcntl" => {
+                if !(args.len() == 2 || args.len() == 3) {
+                    throw_ub_format!("incorrect number of arguments, got {}, needed 2 or 3", args.len());
+                }
```
Comment (RalfJung) [inline src/shims/foreign_items/posix.rs:61]: This seems oddly permissive... it seems better to pass `&args[2..]` to `this.fcntl` and then in there do the proper validation for each possible command.

File: src/shims/foreign_items/posix/linux.rs [hunk 7]
```diff
@@ -81,16 +89,20 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
                     .to_machine_usize(this)?;
 
+                if args.is_empty() {
+                    throw_ub_format!("incorrect number of arguments, needed at least 1");
+                }
                 match this.read_scalar(args[0])?.to_machine_usize(this)? {
                     // `libc::syscall(NR_GETRANDOM, buf.as_mut_ptr(), buf.len(), GRND_NONBLOCK)`
                     // is called if a `HashMap` is created the regular way (e.g. HashMap<K, V>).
                     id if id == sys_getrandom => {
-                        // The first argument is the syscall id, so skip over it.
```
Comment (RalfJung) [inline src/shims/foreign_items/posix/linux.rs:88]: Please keep the comment, it was useful.

File: src/shims/foreign_items/posix.rs (hunk 1 above)
Comment (toc-the-younger) [reply src/shims/foreign_items/posix.rs:281]: That was my original thought reading this issue, and I'd probably prefer that version ultimately. At least one version I've tried managed to ICE, so I don't know if using const generics for this is worth it just yet?

```
// in helpers.rs under
// pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
    fn read_scalars<const N: usize>(&mut self, args: &[OpTy<'tcx, Tag>]) -> InterpResult<'tcx, [Scalar<Tag>; N]> {
        let this = self.eval_context_mut();
        if args.len() != N {
            throw_ub_format!("incorrect number of arguments, got {}, needed {}", args.len(), N);
        }
        let mut scalars = [Scalar::<Tag>::from_u8(0); N];
        for (i, &arg) in args.iter().enumerate() {
            scalars[i] = this.read_scalar(arg)?.not_undef()?;
        }
        Ok(scalars)
    }

// then
--- a/src/shims/intrinsics.rs
+++ b/src/shims/intrinsics.rs
@@ -416,8 +416,10 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx

             "offset" => {
                 check_arg_count(args, 2)?;
-                let offset = this.read_scalar(args[1])?.to_machine_isize(this)?;
-                let ptr = this.read_scalar(args[0])?.not_undef()?;
+                let [offset, ptr] = this.read_scalars::<2usize>(args);
+                let offset = offset.to_machine_isize(this)?;
                 let result_ptr = this.pointer_offset_inbounds(ptr, substs.type_at(0), offset)?;
                 this.write_scalar(result_ptr, dest)?;
             }
```

Comment (RalfJung) [reply src/shims/foreign_items/posix.rs:281]: Yeah, const-generic is pretty ICE-prone... but if we restrict ourselves to the patterns libstd uses, it should not be too bad I hope. We could always fall back to a macro here.

However, please avoid this pattern:
```rust
        let mut scalars = [Scalar::<Tag>::from_u8(0); N];
        for (i, &arg) in args.iter().enumerate() {
            scalars[i] = this.read_scalar(arg)?.not_undef()?;
        }
```
I'd rather not risk accidentally returning the default value. This is exactly what we have [`TryFrom` instances for](https://doc.rust-lang.org/nightly/src/core/array/mod.rs.html#137-147).

Comment (toc-the-younger) [reply src/shims/foreign_items/posix.rs:281]: > However, please avoid this pattern:

I was indeed using TryFrom in other versions of this, but those versions wouldn't compile either. This is merely the version that ICE'd.

Comment (eddyb) [reply src/shims/foreign_items/posix.rs:281]: I think the ICE is from passing `2usize` to a method's const generic parameter (we've been meaning to ban that sort of thing, it's unsupportable).

Can you try to rely on inference instead? It should "just work".

**Comment by eddyb** [comment]
In case it got lost, I left a comment a few days ago: https://github.com/rust-lang/miri/pull/1298#discussion_r406100614.

File: src/shims/intrinsics.rs (hunk 2 above)
Comment (RalfJung) [reply src/shims/intrinsics.rs:389]: I added a test in https://github.com/rust-lang/miri/pull/1337.

File: src/shims/foreign_items/posix.rs (hunk 1 above)
Comment (toc-the-younger) [reply src/shims/foreign_items/posix.rs:281]: @eddyb Tried `this.read_scalars::<2>(args);` and `this.read_scalars(args);`, they did not seem to compile. But I could try again. One question I have is if one of these versions using const generics does manage to work can we just depend on it continuing to work? Or might it start breaking in the future?

Comment (toc-the-younger) [reply src/shims/foreign_items/posix.rs:281]: Update: I've got a version that works, redoing the PR with that.

File: src/helpers.rs (hunk 3 above)
Comment (toc-the-younger) [reply src/helpers.rs]: When I cause this to error I seem to be able to figure out what `link_name` was, but I can still pass it in explicitly if you want.

Comment (toc-the-younger) [reply src/helpers.rs]: Which is to say, should `link_name` be passed in here or just generally be added to any error in `emulate_foreign_item`?

File: src/helpers.rs [hunk 8]
```diff
@@ -471,6 +471,15 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     }
 }
 
+/// Check that the number of args is what we expect.
+pub fn check_args<'a, 'tcx, const N: usize>(args: &'a [OpTy<'tcx, Tag>]) -> InterpResult<'tcx, &'a [OpTy<'tcx, Tag>; N]>
```
Comment (toc-the-younger) [inline src/helpers.rs:475]: I think this function should check more than just arity (e.g. every scalar that goes through here should at least pass`this.read_scalar(ptr)?.not_undef()?`). But probably every argument should be validated to be the correct type as well when possible. Some shims that ignore a flag or mask drop it on the floor, some at least validate that it's the correct type (and then ignore it). That should all probably be a separate issue though.

File: src/helpers.rs [hunk 9]
```diff
@@ -258,17 +258,17 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         /// Visiting the memory covered by a `MemPlace`, being aware of
         /// whether we are inside an `UnsafeCell` or not.
         struct UnsafeCellVisitor<'ecx, 'mir, 'tcx, F>
-        where
-            F: FnMut(MPlaceTy<'tcx, Tag>) -> InterpResult<'tcx>,
+            where
```
Comment (RalfJung) [inline src/helpers.rs:261]: Looks like we got a bunch of unrelated changes here?

File: src/helpers.rs (hunk 8 above)
Comment (RalfJung) [reply src/helpers.rs:475]: Almost all shims use their argument, so I am not sure if it's worth repeating that information by forcing them to give an explicit signature here or so.

Note that not all arguments have to be scalars, so even `read_scalar` is going too far.

IMO we can leave it to each shim to validate the size of its arguments, to whatever extend is appropriate. Anyway UB detection for shims (this is basically library UB, not language UB) is on a best-effort basis.

Comment (RalfJung) [inline src/helpers.rs:475]: The function name should reflect that this is just about the argument count... maybe `check_arg_count`?

File: src/lib.rs [hunk 10]
```diff
@@ -7,6 +7,9 @@
 #![warn(rust_2018_idioms)]
 #![allow(clippy::cast_lossless)]
 
+#![allow(incomplete_features)]
```
Comment (RalfJung) [inline src/lib.rs:10]: Does this work without the `!` as well?

File: src/shims/foreign_items.rs [hunk 11]
```diff
@@ -197,25 +199,29 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
-                let size = this.read_scalar(args[0])?.to_machine_usize(this)?;
+                let &[size] = check_args(args)?;
+                let size = this.read_scalar(size)?.to_machine_usize(this)?;
                 let res = this.malloc(size, /*zero_init:*/ false, MiriMemoryKind::C);
                 this.write_scalar(res, dest)?;
             }
             "calloc" => {
-                let items = this.read_scalar(args[0])?.to_machine_usize(this)?;
-                let len = this.read_scalar(args[1])?.to_machine_usize(this)?;
+                let &[items, len] = check_args(args)?;
```
Comment (RalfJung) [inline src/shims/foreign_items.rs:208]: This looks really nice. :)

**Comment by RalfJung** [comment]
Could you add a testcase for some intrinsic and some foreign item that we call with too few / too many arguments? (There's no point in testing *all* of them, but we should at least cover one, I think.)

File: src/helpers.rs (hunk 3 above)
Comment (RalfJung) [reply src/helpers.rs]: Hm... can you show an example for how the error looks? You are right we don't print the function name for other errors like bad argument sizes, so it's probably not worth the effort.

File: src/shims/foreign_items/posix.rs [hunk 12]
```diff
@@ -23,43 +24,58 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
             "fcntl" => {
-                let result = this.fcntl(args[0], args[1], args.get(2).cloned())?;
+                let result = if args.len() == 2 {
+                    let &[fd, cmd] = check_args(args)?;
+                    this.fcntl(fd, cmd, None)?
+                } else {
+                    // If args.len() isn't 2 or 3 this will error appropriately.
+                    let &[fd, cmd, start] = check_args(args)?;
+                    this.fcntl(fd, cmd, Some(start))?
+                };
```
Comment (RalfJung) [inline src/shims/foreign_items/posix.rs:66]: I think it would be better to just pass `args` to `fcntl`, and let it do the checking.

File: src/shims/foreign_items/posix/linux.rs [hunk 13]
```diff
@@ -143,12 +157,13 @@ fn getrandom<'tcx>(
     args: &[OpTy<'tcx, Tag>],
     dest: PlaceTy<'tcx, Tag>,
 ) -> InterpResult<'tcx> {
-    let ptr = this.read_scalar(args[0])?.not_undef()?;
-    let len = this.read_scalar(args[1])?.to_machine_usize(this)?;
+    let &[ptr, len, _flags] = check_args(args)?;
```
Comment (RalfJung) [inline src/shims/foreign_items/posix/linux.rs:160]: This will, however, show a wrong error... namely if you pass 3 arguments it will show "passed 2 arguments expected 3", whereas really it was "passed 3 arguments expected 4".

Probably it is better to change the type signature here to just take 3 arguments fixed, and let the caller do `check_args`.

File: src/shims/foreign_items/posix/macos.rs [hunk 14]
```diff
@@ -15,85 +16,102 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
-                let data = this.read_scalar(args[1])?.not_undef()?;
+                let data = this.read_scalar(data)?.not_undef()?;
                 let active_thread = this.get_active_thread()?;
                 this.machine.tls.set_macos_thread_dtor(active_thread, dtor, data)?;
             }
 
             // Querying system information
             "pthread_get_stackaddr_np" => {
-                let _thread = this.read_scalar(args[0])?.not_undef()?;
+                let &[_thread] = check_args(args)?;
```
Comment (RalfJung) [inline src/shims/foreign_items/posix/macos.rs:107]: The `_thread` is not unused... I think it's better not to start it with an `_` then. That way we get a warning if we fail to at least do a basic `read_scalar` on it.

**Comment by RalfJung** [comment]
Also please rebase, we prefer not to have merge conflicts in PRs.

File: src/lib.rs (hunk 10 above)
Comment (toc-the-younger) [reply src/lib.rs:10]: Not directly:
```
error: an inner attribute is not permitted following an outer attribute
  --> src/lib.rs:11:1
   |
10 | #[allow(incomplete_features)]
   | ----------------------------- previous outer attribute
11 | #![feature(const_generics)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ not permitted following an outer attribute
   |
   = note: inner attributes, like `#![no_std]`, annotate the item enclosing them, and are usually found at the beginning of source files. Outer attributes, like `#[test]`, annotate the item following them.
```

Comment (RalfJung) [reply src/lib.rs:10]: Okay. Well seems fine then.

File: tests/compile-fail/check_arg_count.rs [hunk 15]
```diff
@@ -0,0 +1,7 @@
+#![feature(core_intrinsics)]
+
+use std::intrinsics;
+
+fn main() {
+    unsafe { intrinsics::forget(); } //~ ERROR this function takes 1 argument but 0 arguments were supplied
+}
```
Comment (RalfJung) [inline tests/compile-fail/check_arg_count.rs:7]: Please also have a test for passing *too many* arguments.

File: src/helpers.rs (hunk 3 above)
Comment (toc-the-younger) [reply src/helpers.rs]: When I change it in the tests I get something like
```
error: Undefined Behavior: incorrect number of arguments, got 1, needed 0
  --> $DIR/intrinsics.rs:27:14
   |
27 |     unsafe { intrinsics::forget(); }
   |              ^^^^^^^^^^^^^^^^^^^^^^^^ incorrect number of arguments, got 0, needed 1
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
```
along with a bunch of other output.

Comment (RalfJung) [reply src/helpers.rs]: Okay, so it should be pretty clear which function it is talking about. Thanks!

File: src/helpers.rs [hunk 16]
```diff
@@ -471,6 +471,15 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     }
 }
 
+/// Check that the number of args is what we expect.
+pub fn check_arg_count<'a, 'tcx, const N: usize>(args: &'a [OpTy<'tcx, Tag>]) -> InterpResult<'tcx, &'a [OpTy<'tcx, Tag>; N]>
+    where &'a [OpTy<'tcx, Tag>; N]: TryFrom<&'a [OpTy<'tcx, Tag>]> {
+    if let Ok(ops) = args.try_into() {
+        return Ok(ops);
+    }
+    throw_ub_format!("incorrect number of arguments, got {}, needed {}", args.len(), N)
```
Comment (RalfJung) [inline src/helpers.rs:480]:
```suggestion
    throw_ub_format!("incorrect number of arguments: got {}, expected {}", args.len(), N)
```
That's more like the wording we are usually using.

File: src/shims/foreign_items.rs [hunk 17]
```diff
@@ -415,9 +433,10 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
             => {
+                let &[x, exp] = check_arg_count(args)?;
                 // For radix-2 (binary) systems, `ldexp` and `scalbn` are the same.
-                let x = this.read_scalar(args[0])?.to_f64()?;
-                let exp = this.read_scalar(args[1])?.to_i32()?;
+                let x = this.read_scalar(x)?.to_f64()?;
+                let exp = this.read_scalar(exp)?.to_i32()?;
 
                 // Saturating cast to i16. Even those are outside the valid exponent range to
                 // `scalbn` below will do its over/underflow handling.
```
Comment (RalfJung) [inline src/shims/foreign_items.rs:442]: There is one more below, `llvm.x86.sse2.pause`, which should be checked to not have any arguments.

File: src/shims/foreign_items/posix/linux.rs [hunk 18]
```diff
@@ -92,18 +101,23 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                     .eval_libc("SYS_statx")?
                     .to_machine_usize(this)?;
 
+                if args.is_empty() {
+                    throw_ub_format!("incorrect number of arguments for syscall, needed at least 1");
```
Comment (RalfJung) [inline src/shims/foreign_items/posix/linux.rs:105]:
```suggestion
                    throw_ub_format!("incorrect number of arguments for syscall: got 0, expected at least 1");
```

File: src/shims/foreign_items/windows.rs [hunk 19]
```diff
@@ -88,41 +97,48 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+                let size = this.read_scalar(size)?.to_machine_usize(this)?;
                 let zero_init = (flags & 0x00000008) != 0; // HEAP_ZERO_MEMORY
                 let res = this.malloc(size, zero_init, MiriMemoryKind::WinHeap);
                 this.write_scalar(res, dest)?;
             }
             "HeapFree" => {
-                let _handle = this.read_scalar(args[0])?.to_machine_isize(this)?;
-                let _flags = this.read_scalar(args[1])?.to_u32()?;
-                let ptr = this.read_scalar(args[2])?.not_undef()?;
+                let &[handle, _flags, ptr] = check_arg_count(args)?;
```
Comment (RalfJung) [inline src/shims/foreign_items/windows.rs:109]: Please rename `_flags` -> `flags` to make sure we validate it (as we do below).

File: src/shims/foreign_items/windows.rs [hunk 20]
```diff
@@ -88,41 +97,48 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+                let ptr = this.read_scalar(ptr)?.not_undef()?;
                 this.free(ptr, MiriMemoryKind::WinHeap)?;
                 this.write_scalar(Scalar::from_i32(1), dest)?;
             }
             "HeapReAlloc" => {
-                let _handle = this.read_scalar(args[0])?.to_machine_isize(this)?;
-                let _flags = this.read_scalar(args[1])?.to_u32()?;
-                let ptr = this.read_scalar(args[2])?.not_undef()?;
-                let size = this.read_scalar(args[3])?.to_machine_usize(this)?;
+                let &[handle, _flags, ptr, size] = check_arg_count(args)?;
```
Comment (RalfJung) [inline src/shims/foreign_items/windows.rs:117]: Same, please rename `_flags` -> `flags`.

File: src/shims/foreign_items/posix/linux.rs [hunk 21]
```diff
@@ -112,12 +126,14 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
-                let _pid = this.read_scalar(args[0])?.to_i32()?;
-                let _cpusetsize = this.read_scalar(args[1])?.to_machine_usize(this)?;
-                let _mask = this.deref_operand(args[2])?;
+                let &[pid, cpusetsize, mask] = check_arg_count(args)?;
+                this.read_scalar(pid)?.to_i32()?;
+                this.read_scalar(cpusetsize)?.to_machine_usize(this)?;
+                this.deref_operand(mask)?;
                 // FIXME: we just return an error; `num_cpus` then falls back to `sysconf`.
                 let einval = this.eval_libc("EINVAL")?;
                 this.set_last_error(einval)?;
```
Comment (RalfJung) [inline src/shims/foreign_items/posix/linux.rs:139]: Could you also add a check to `pthread_getattr_np` below?

File: src/shims/fs.rs [hunk 22]
```diff
@@ -322,16 +322,22 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
-        start_op: Option<OpTy<'tcx, Tag>>,
+        args: &[OpTy<'tcx, Tag>],
     ) -> InterpResult<'tcx, i32> {
         let this = self.eval_context_mut();
 
         this.check_no_isolation("fcntl")?;
 
-        let fd = this.read_scalar(fd_op)?.to_i32()?;
-        let cmd = this.read_scalar(cmd_op)?.to_i32()?;
+        let (fd, cmd, start) = if args.len() == 2 {
```
Comment (RalfJung) [inline src/shims/fs.rs:331]: This seems odd... I think this should look more like what you did for `syscall`: first just extract `cmd`, then in the `cmd` dispatch below extract the right number of arguments.

Comment (RalfJung) [reply src/shims/fs.rs:331]: Ah, this one has *two* mandatory arguments. So first extract `fd` and `cmd`, and then get the rest in the dispatch below.

File: tests/compile-fail/check_arg_count.rs [hunk 23]
```diff
@@ -0,0 +1,8 @@
+#![feature(core_intrinsics)]
+
+use std::intrinsics;
+
+fn main() {
+    unsafe { intrinsics::forget(); } //~ ERROR this function takes 1 argument but 0 arguments were supplied
+    unsafe { intrinsics::forget(1, 2); } //~ ERROR this function takes 1 argument but 2 arguments were supplied
```
Comment (RalfJung) [inline tests/compile-fail/check_arg_count.rs:7]: This won't work, we stop execution after the first failure.  You have to write two tests, I am afraid.

Comment (RalfJung) [reply tests/compile-fail/check_arg_count.rs:7]: It seems like CI passes; I have no idea how that is possible... I'll need to investigate that. But please still make it two tests.

Comment (RalfJung) [reply tests/compile-fail/check_arg_count.rs:7]: Wow, something really strange is happening... I am seeing both errors with `./miri run`, whereas usually we abort on the first sight of UB...

Comment (RalfJung) [reply tests/compile-fail/check_arg_count.rs:7]: Oh lol. You are testing the wrong errors. :rofl: 

These errors are from the type checker. You will need to define your own wrong signature for the intrinsic and call it like that for Miri to even start doing something.

Probably it is better to do that with a foreign function than an intrinsic. See `shim_arg_size` for an example test that does something like that.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/miri/pull/1390 added a new shim which also needs to be added here.

**Comment by RalfJung** [comment]
This looks great, thanks a lot. :-)
@bors r+
